### PR TITLE
MNTOR-1800: unsubscribe from free monthly report

### PR DESCRIPTION
**Pull request by mansaj**
# References:

Jira: MNTOR-3582

# Description
1. Util function for table `subscriber_email_preferences`
The util funcs take care of the fetches and updates of email preferences. It gets the stored settings from 2 tables and consolidate them. Follow up work will be tracked in MNTOR-3557

2. Util function to generate unsubscribe URL
3. API endpoint to verify unsub url and unsub the user from monthly monitor reports

# Screenshot (if applicable)

Not applicable.

# How to test
* get the output of `console.info(generateUnsubscribeLink(session?.user.email || ""));` on any page
* visit that link
  * if the pref table is empty, a row will be created
  * if the pref table record already exists, it will check for the value of `monthly_monitor_report_free` value and change it to false if not

# Checklist (Definition of Done)
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [x] I've added or updated the relevant sections in readme and/or code comments
- [x] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [x] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: src/app/api/utils/email.tsx [hunk 1]
```diff
@@ -54,3 +61,45 @@ export async function sendVerificationEmail(
     ),
   );
 }
+
+export async function generateUnsubscribeLinkForSubscriber(
```
Comment (mansaj) [inline src/app/api/utils/email.tsx:65]: @rhelmer changed this function to always generate a fresh token and update the db (upsert a new row in the new table for subscriber)

**Review by rhelmer — CHANGES_REQUESTED** [review]
lgtm and wfm locally... I was going to say `subscriber_email_preferences.primary_email` should have a FK constraint on `email_addresses` or `subscriber.primary_email`, but I guess that we want to keep a record permanently of the email address that the user opted out of?

What happens if a user changes their primary email address, will they start getting these emails again?

**Review by rhelmer — APPROVED** [review]

File: src/app/api/utils/email.tsx [hunk 2]
```diff
@@ -54,3 +61,48 @@ export async function sendVerificationEmail(
...
+  try {
+    const unsubToken = randomString();
+    const sub = await addEmailPreferenceForSubscriber(
+      subscriber.id,
+      {
+        unsubscribe_token: unsubToken,
+      },
+      ["unsubscribe_token"],
+    );
+    return `${process.env.SERVER_URL}/api/v1/user/unsubscribe-email?email=${subscriber.primary_email}&token=${sub.unsubscribe_token}`;
```
[the remark this answers is not in the document]
Comment (mansaj) [reply src/app/api/utils/email.tsx:77]: @Vinnl 
1. yes, unique per use
2. what would the URL be? If I understand correctly, we'd be able to get the email address directly from the email itself (when it's generated) and I'll just pass the `token=` only then

[the remark this answers is not in the document]
Comment (mansaj) [reply src/app/api/utils/email.tsx:77]: sounds good

[the remark this answers is not in the document]
Comment (Vinnl) [reply src/app/api/utils/email.tsx]: @mansaj That does require the API also just accepting only the token - see a suggestion in https://github.com/mozilla/blurts-server/pull/5019.

One other thing I noticed: all the code in this PR seems to assume that we'll be unsubscribing either the monthly free or the monthly Plus report, but it is expected that there will be more emails that the user might want to unsubscribe from in the future. Should we not e.g. add another parameter that indicates the email to unsubscribe from?

File: src/db/tables/subscriber_email_preferences.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+      .insert({
+        subscriber_id: subscriberId,
+        unsubscribe_token: preference.unsubscribe_token || "",
+        monthly_monitor_report_free:
+          preference.monthly_monitor_report_free ?? true,
+        monthly_monitor_report_free_at:
+          preference.monthly_monitor_report_free_at ?? null,
+      })
+      .onConflict("subscriber_id")
+      .merge(fieldsToOverwrite)
```
Comment (Vinnl) [inline src/db/tables/subscriber_email_preferences.ts:58]: Oh also, I called this function to be able to test out unsubscription, and it threw an error unless I commented out these two lines. Possibly due to passing an empty array as the third argument, in which case: maybe some JSDoc explaining what it's supposed to be used for could be helpful? It wasn't clear to me 😅

File: src/app/api/utils/email.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (mansaj) [reply src/app/api/utils/email.tsx]: It's unclear to me whether per email unsubscribe is on the roadmap. In order to minimize logging PII, only token is fine for now.

If there's a prod need to rework unsubscribe, we might have to use something like a subscriber ID Or email ID to indicate which email to unsub from

File: src/db/tables/subscriber_email_preferences.ts (hunk 3 above)
Comment (mansaj) [reply src/db/tables/subscriber_email_preferences.ts:61]: added a note

**Pull request merged**